File: Cpp/355_Design_Twitter.cpp
class Twitter {
public:
    /** Initialize your data structure here. */
    Twitter() : time(0) {
        
    }
    
    /** Compose a new tweet. */
    void postTweet(int userId, int tweetId) {
        tweet_lists[userId].push_back(Tweet(tweetId, time++));
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    vector<int> getNewsFeed(int userId) {
        vector<int> result;
        auto comp = [](const Tweet &t1, const Tweet &t2) {
            return t1.time < t2.time;
        };
        priority_queue<Tweet, vector<Tweet>, decltype(comp)> pq(comp);
        
        vector<Tweet>& t = tweet_lists[userId];
        int t_size = t.size();
        for(int i=t.size()-1; i>=0 && i>=t_size-10; --i){
            pq.push(t[i]);
        }
        
        for(const auto& follower_id : followers[userId]){
            vector<Tweet>& t = tweet_lists[follower_id];
            int t_size = t.size();
            for(int i=t.size()-1; i>=0 && i>=t_size-10; --i){
                pq.push(t[i]);
            }
        }
        
        while(!pq.empty() && result.size() < 10){
            result.push_back(pq.top().id);
            pq.pop();
        }
        
        return result;
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    void follow(int followerId, int followeeId) {
        if(followerId == followeeId) return;
        followers[followerId].insert(followeeId);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    void unfollow(int followerId, int followeeId) {
        followers[followerId].erase(followeeId);
    }

private:
struct Tweet{
    int id;
    int time;
    
    Tweet(int id, int time) : id(id), time(time) {
    }
};
    
    unordered_map<int, unordered_set<int>> followers;
    unordered_map<int, vector<Tweet>> tweet_lists;
    int time;
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */

File: Kotlin/355_Design_Twitter.kt
class Twitter() {
    val followers = mutableMapOf<Int, MutableSet<Int>>()
    val tweetList = mutableMapOf<Int, MutableList<Tweet>>()
    var time = 0

    fun postTweet(userId: Int, tweetId: Int) {
        tweetList[userId] = tweetList[userId] ?: mutableListOf()
        tweetList[userId]?.add(Tweet(tweetId, time++))
    }

    fun getNewsFeed(userId: Int): List<Int> {
        val result = mutableListOf<Int>()
        val pq = PriorityQueue<Tweet>(Comparator<Tweet>{t1, t2 -> t2.time - t1.time})
        
        var tweet = tweetList[userId] ?: mutableListOf()
        var tweetSize = tweet?.size ?: 0
        for(i in tweetSize - 1 downTo maxOf(0, tweetSize - 10)) {
            pq.add(tweet[i]!!)
        }
        
        followers[userId]?.forEach{ followerId ->
            tweet = tweetList[followerId] ?: mutableListOf()
            tweetSize = tweet?.size ?: 0
            for(i in tweetSize - 1 downTo maxOf(0, tweetSize - 10)) {
                pq.add(tweet[i]!!)
            }
        }
        
        while(!pq.isEmpty() && result.size < 10){
            result.add(pq.peek().id)
            pq.poll()
        }

        return result
    }

    fun follow(followerId: Int, followeeId: Int) {
        if (followeeId == followerId) return
        followers[followerId] = followers[followerId] ?: mutableSetOf()
        followers[followerId]?.add(followeeId)
    }

    fun unfollow(followerId: Int, followeeId: Int) {
        followers[followerId]?.remove(followeeId)
    }

}

data class Tweet(val id: Int, val time: Int)

/**
 * Your Twitter object will be instantiated and called as such:
 * var obj = Twitter()
 * obj.postTweet(userId,tweetId)
 * var param_2 = obj.getNewsFeed(userId)
 * obj.follow(followerId,followeeId)
 * obj.unfollow(followerId,followeeId)
 */